[Woo POS] Variation pagination

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #14696
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR adds pagination support for variations in POS.

The feature works fully, but I think we can iterate on it a little in a future PR. There's a lot of duplication between the initial sync and next page sync – from a quick experiment, we could remove the initial sync and rely on defaults, especially since there's a separate resync function for PTR.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

Open the POS, and select a variable product to open.
Note that the first page is loaded, and if you have enough variations, scrolling down will load subsequent pages until all variations are loaded.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

N.B. The page size is still 100 in this PR; in #14843 it will be reduced to 25.

For testing, I recommend reducing it to 15 so you can see more pages loaded. If you choose too small a number that doesn't fill the screen, the scroll observation to trigger next page load won't work.

Additionally, variations will only show up if you're using WC 9.6, which is not yet released. You can work around this by commenting out ProductsRemote line 214, so that the root products list is not filtered to simple products only.

    //            ParameterKey.productType: POSConstants.productType,

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/user-attachments/assets/9c30f5c8-b0e5-464b-b4fb-554c524298f1


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

joshheald committed:
14696 Use baseItem to load item lists

joshheald committed:
14696 Add page info to variation fetches

joshheald committed:
14696 Load next variation page on child item lists

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
14696 Extract child item fetch and futureproof it

The switch statement now requires that newly added POSItems are specifically handled.

jaclync commented on Yosemite/Yosemite/PointOfSale/PointOfSaleItemService.swift:
```diff
@@ -85,7 +86,7 @@ public final class PointOfSaleItemService: PointOfSaleItemServiceProtocol {
                                      variationID: variation.productVariationID))
             }),
             // TODO-14696: pagination support for variations lists
```
nit: we can remove this TODO comment now

jaclync replied:
Update: removed in https://github.com/woocommerce/woocommerce-ios/pull/14849/commits/3b339193ad551edbdd26b701187436fd578f995d.

jaclync committed:
Remove outdated TODO comment now that pagination support is implemented.

jaclync commented:
Works great! :shipit: I noticed an infinite scroll issue from the trigger state which I will fix separately.

> The feature works fully, but I think we can iterate on it a little in a future PR. There's a lot of duplication between the initial sync and next page sync – from a quick experiment, we could remove the initial sync and rely on defaults, especially since there's a separate resync function for PTR.

I agree, the initial load could just use the PTR resync IMO. The difference between initial load and reload seems to be just resetting the internal states in the reload, which works for the initial load as well. I will try including this in the PTR work next.
